stdlib/sql/src/main/java/org/ballerinalang/sql/datasource/SQLDatasource.java

```diff
@@ -0,0 +1,241 @@
+                    config.addDataSourceProperty(Constants.Options.URL, sqlDatasourceParams.url);
+                }
+            }
+            config.setDataSourceClassName(sqlDatasourceParams.datasourceName);
+            if (sqlDatasourceParams.connectionPool != null) {
+                int maxOpenConn = sqlDatasourceParams.connectionPool.
+                        getIntValue(Constants.ConnectionPool.MAX_OPEN_CONNECTIONS).intValue();
+                if (maxOpenConn < 0) {
+                    config.setMaximumPoolSize(maxOpenConn);
+                }
+
+                Object connLifeTimeSec = sqlDatasourceParams.connectionPool
+                        .get(Constants.ConnectionPool.MAX_CONNECTION_LIFE_TIME_SECONDS);
+                if (connLifeTimeSec instanceof DecimalValue) {
+                    DecimalValue connLifeTime = (DecimalValue) connLifeTimeSec;
+                    if (connLifeTime.floatValue() > 0) {
+                        long connLifeTimeMS = Double.valueOf(connLifeTime.floatValue() * 1000).longValue();
+                        config.setMaxLifetime(connLifeTimeMS);
+                    }
+                }
+                int minIdleConnections = sqlDatasourceParams.connectionPool
+                        .getIntValue(Constants.ConnectionPool.MIN_IDLE_CONNECTIONS).intValue();
+                if (minIdleConnections < 0) {
+                    config.setMinimumIdle(minIdleConnections);
+                }
+            }
+            if (sqlDatasourceParams.options != null) {
+                MapValue<String, Object> optionMap = (MapValue<String, Object>) sqlDatasourceParams.options;
+                optionMap.entrySet().forEach(entry -> {
+                    if (SQLDatasourceUtils.isSupportedDbOptionType(entry.getValue())) {
+                        config.addDataSourceProperty(entry.getKey(), entry.getValue());
+                    } else {
+                        throw ErrorGenerator.getSQLApplicationError("unsupported type " + entry.getKey()
+                                + " for the db option");
+                    }
+                });
+            }
+            hikariDataSource = new HikariDataSource(config);
+            Runtime.getRuntime().addShutdownHook(new Thread(this::closeConnectionPool));
+        } catch (Throwable t) {
+            String message = "error in sql connector configuration: " + t.getMessage();
+            if (t.getCause() != null) {
+                message += ":" + t.getCause().getMessage();
+            }
+            throw ErrorGenerator.getSQLApplicationError(message);
+        }
+    }
+    /**
+     * This class encapsulates the parameters required for the initialization of {@code SQLDatasource} class.
+     */
+    public static class SQLDatasourceParams {
+        private String url;
+        private String user;
+        private String password;
+        private String datasourceName;
+        private MapValue connectionPool;
```

Comment:
Instead of using raw `MapValue` shall we use the proper types? ex: `MapValue<String, Object>`

Reply:
The MapValue interface considers the key as String, and hence I don't think there is any difference of using the raw MapValue or MapValue<String, Object> here.